KubeArmor/core/kubeUpdate.go

```diff
@@ -1070,26 +1076,40 @@ func (dm *KubeArmorDaemon) UpdateSecurityPolicy(action string, secPolicyType str
 
 		// update a security policy
 		if secPolicyType == KubeArmorPolicy {
-			if kl.MatchIdentities(secPolicy.Spec.Selector.Identities, endPoint.Identities) && (len(secPolicy.Spec.Selector.Containers) == 0 || kl.ContainsElement(secPolicy.Spec.Selector.Containers, endPoint.ContainerName)) {
+			if len(secPolicy.Spec.Selector.Containers) == 0 || kl.ContainsElement(secPolicy.Spec.Selector.Containers, endPoint.ContainerName) {
 				if action == "ADDED" {
-					// add a new security policy if it doesn't exist
-					new := true
-					for _, policy := range endPoint.SecurityPolicies {
-						if policy.Metadata["namespaceName"] == secPolicy.Metadata["namespaceName"] && policy.Metadata["policyName"] == secPolicy.Metadata["policyName"] {
-							new = false
-							break
+					if kl.MatchIdentities(secPolicy.Spec.Selector.Identities, endPoint.Identities) && kl.MatchExpIdentities(secPolicy.Spec.Selector, endPoint.Identities) {
+						// add a new security policy if it doesn't exist
+						new := true
+						for _, policy := range endPoint.SecurityPolicies {
+							if policy.Metadata["namespaceName"] == secPolicy.Metadata["namespaceName"] && policy.Metadata["policyName"] == secPolicy.Metadata["policyName"] {
+								new = false
+								break
+							}
+						}
+						if new {
+							endPoint.SecurityPolicies = append(endPoint.SecurityPolicies, secPolicy)
 						}
-					}
-					if new {
-						endPoint.SecurityPolicies = append(endPoint.SecurityPolicies, secPolicy)
 					}
 				} else if action == "MODIFIED" {
+					// in case new labels are added in the policy, check if identities match, if yes, add policy in endPoint
```

Comment:
it's important to document this behaviour, because when target workloads changed by new selectors it is natural to assume that policy is no more valid for previously targeted workloads but the rules will not be reverted for those workloads right! just that if any rules updated those will not be valid for those workloads right!